core/src/com/cloud/storage/template/S3TemplateDownloader.java

```diff
@@ -19,303 +19,235 @@
 
-        try {
-            request = new GetMethod(downloadUrl);
-            request.getParams().setParameter(HttpMethodParams.RETRY_HANDLER, myretryhandler);
-            completionCallback = callback;
-
-            Pair<String, Integer> hostAndPort = UriUtils.validateUrl(downloadUrl);
-            fileName = StringUtils.substringAfterLast(downloadUrl, "/");
-            fileExtension = StringUtils.substringAfterLast(fileName, ".");
-
-            if (proxy != null) {
-                client.getHostConfiguration().setProxy(proxy.getHost(), proxy.getPort());
-                if (proxy.getUserName() != null) {
-                    Credentials proxyCreds = new UsernamePasswordCredentials(proxy.getUserName(), proxy.getPassword());
-                    client.getState().setProxyCredentials(AuthScope.ANY, proxyCreds);
-                }
-            }
-            if ((user != null) && (password != null)) {
-                client.getParams().setAuthenticationPreemptive(true);
-                Credentials defaultcreds = new UsernamePasswordCredentials(user, password);
-                client.getState().setCredentials(
-                        new AuthScope(hostAndPort.first(), hostAndPort.second(), AuthScope.ANY_REALM), defaultcreds);
-                s_logger.info("Added username=" + user + ", password=" + password + "for host " + hostAndPort.first()
-                        + ":" + hostAndPort.second());
-            } else {
-                s_logger.info("No credentials configured for host=" + hostAndPort.first() + ":" + hostAndPort.second());
-            }
-        } catch (IllegalArgumentException iae) {
-            errorString = iae.getMessage();
-            status = TemplateDownloader.Status.UNRECOVERABLE_ERROR;
-            inited = false;
-        } catch (Exception ex) {
-            errorString = "Unable to start download -- check url? ";
-            status = TemplateDownloader.Status.UNRECOVERABLE_ERROR;
-            s_logger.warn("Exception in constructor -- " + ex.toString());
-        } catch (Throwable th) {
-            s_logger.warn("throwable caught ", th);
-        }
+        // Follow redirects
+        this.getMethod.setFollowRedirects(true);
+
+        // Set file extension.
+        this.fileExtension = StringUtils.substringAfterLast(StringUtils.substringAfterLast(downloadUrl, "/"), ".");
+
+        // Calculate and set S3 Key.
+        this.s3Key = join(asList(installPath, StringUtils.substringAfterLast(downloadUrl, "/")), S3Utils.SEPARATOR);
+
+        // Set proxy if available.
+        HTTPUtils.setProxy(proxy, this.httpClient);
+
+        // Set credentials if available.
+        HTTPUtils.setCredentials(username, password, this.httpClient);
     }
 
     @Override
     public long download(boolean resume, DownloadCompleteCallback callback) {
-        switch (status) {
-        case ABORTED:
-        case UNRECOVERABLE_ERROR:
-        case DOWNLOAD_FINISHED:
+        if (!status.equals(Status.NOT_STARTED)) {
```

Comment:
If you want, after this PR is merged, please come to SBP and the two of us can have a Refactoring Hackathon and make this better. It's fine by now, but we could gain some more readability and extensibility if we decouple part of the code. The status, for example, should be in the Callbacl that is in the parameter list. It could also be used to delated to an adapter which action to take once the callback returns.

With only 2 small changes we could get rid of all those IFs and anonymous classes.